### PR TITLE
New module gmtbinstats for statistical summaries inside circles, hexagons, rectangles on a grid

### PR DESCRIPTION
**Pull request by PaulWessel**
**Description of proposed changes**

This new module **gmtbinstats** follows the suggestion in #3942 by implementing those ideas.  It offers 17 different operators and most of them can be used with or without weights.  The results can be normalized by the search circle area for density values.  I have documented it and added two test scripts that run most of the operations.  Attached is the result of the unweighted result by gridding the populations of the World's capitals:

nonweighted

The tags referes to the -Ccode used, with (**a** = mean, **d** = MAD, **s** = stdev, **i** = interquartile range, **l** = minimum, **m** -= median, **n** = count, **o** = LMS scale, **p** = mode, **q75** the 75% quartile (any quantile is allowed), **g** = full range (max-min), **z** = sum.

It was quick to put together since it inherited most of the i/o machinery from **nearneighbor** and most of the algorithms are inherited from **gmtmath** and **grdmath**, so it is mostly tested indirectly.  Looking at the results the relationship between the various maps seems sensible.

### CODE REVIEW

**Comment by PaulWessel** [comment]
Some users already giving the new module a [spin](https://forum.generic-mapping-tools.org/t/number-of-datapoints-area/765/11).

**Comment by PaulWessel** [comment]
Added support for hexagonal binning via **-T.**  This requires **-I** to set the y-increment and then dx is set to sqrt(3) * dy as required by the geometry.  We may adjust xmax for the region to be inclusive.  Below is what it may look like, and the yellow data points are properly assigned to the correct hexagons.  Two new tests were added to demonstrate the hexagonal tiling.

hexgrid

File: src/gmtcount.c [hunk 1]
```diff
@@ -0,0 +1,935 @@
...
+		case GMTCOUNT_SUM: case GMTCOUNT_COUNTW:
+			G->data[ij] += in[GMT_Z];
+			break;
+		default:	/* count */
+			G->data[ij] += 1.0;
+			break;
+	}
+}
+
+bool outside_hexagon (struct GMT_CTRL *GMT, double x0, double y0, double x, double y, double distance, double radius_inner) {
```
Comment (joa-quim) [inline src/gmtcount.c:381]: This is a bool function returning an int. I had recently hard to find bug in a similar situation and VS2019.

The binning could perhaps be speed up if points are first checked to fall inside the circumscribed rectangle. Also remember that each **GMT_Report()** is an *if* branch, and there seems to be one for each input data point.

Comment (PaulWessel) [reply src/gmtcount.c:381]: OK, fixing the bool.  The GMT_Report was mostly there for me during debugging.  I can place them under #ifdef DEBUG since just there for keeping.
I will look at **gmtselect** for the pre-circle rectangular skip.  But implementing -Tr for rectangles anyway.  For geographic we can do a check on latitude, but longitude by itself cannot work with -S.

Comment (PaulWessel) [reply src/gmtcount.c:381]: latitude checks for rectangle is also tricky across poles.

**Comment by PaulWessel** [comment]
Hi @joa-quim and @seisman.  I think this module is ready to be introduced.  Ideally, you'd try it on some other data to try to break it.

File: doc/rst/source/gmtbinstats.rst [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+**-S**\ *search_radius*
+    Sets the *search_radius* that determines which data points are
+    considered close to a node. Append the distance unit (see `Units`_).
+    Not compatible with **-T**.
+
+.. _-T:
+
+**-T**\ [**h**\|\ **r**]
+    Instead of circular, possibly overlapping areas, select non-overlapping tiling.  Choose between
+    **r**\ ectangularor **h**\ exagonal binning. For **-Tr**, set bin sizes via **-I** and we write
```
Comment (weiji14) [inline doc/rst/source/gmtbinstats.rst:113]:
```suggestion
    **r**\ ectangular **h**\ exagonal binning. For **-Tr**, set bin sizes via **-I** and we write
```

File: doc/rst/source/supplements/module_supplements_purpose.rst_ [hunk 3]
```diff
@@ -30,6 +30,12 @@
 
 .. |mgd77track_purpose| replace:: Plot track-lines of MGD77 cruises
 
+.. |fzanalyzer_purpose| replace:: Analysis of fracture zones using crossing profiles
+
+.. |fzblender_purpose| replace:: Produce a smooth blended FZ trace
+
+.. |mlconverter_purpose| replace:: Convert chrons to ages using selected magnetic timescale
+
```
Comment (weiji14) [inline doc/rst/source/supplements/module_supplements_purpose.rst_:38]: These seem unrelated?

File: doc/rst/source/gmtbinstats.rst [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+-----------
+
+**gmtbinstats** reads arbitrarily located (x,y[,z][,w]) points
+(2-4 columns) from standard input [or *table*] and for each
+node in the specified grid layout determines which points are
+within the given radius.  These point are then used in the
+calculation of the specified statistic. The results may be
+presented as is or may be normalized by the circle area to
+perhaps give density estimates.  Alternatively, select
+hexagonal tiling instead of a rectangular grid layout.
```
Comment (weiji14) [inline doc/rst/source/gmtbinstats.rst:50]:
```suggestion
presented as is or may be normalized by the circle area to
perhaps give density estimates.  Alternatively, select
hexagonal tiling instead or a rectangular grid layout.
```

Not sure if I understand this sentence correctly. Is the circle area based on a rectangular grid, and there is an alternative hexagonal tiling scheme? Or are there circular 'tiles', hexagonal tiles, and rectangular grid tiles.

**Review by weiji14 — COMMENTED** [review]
Exciting stuff! I haven't got time to test this out now, but look forward to trying it once I have time! Just spotted a few typos to fix.

File: doc/rst/source/supplements/module_supplements_purpose.rst_ (hunk 3 above)
Comment (PaulWessel) [reply doc/rst/source/supplements/module_supplements_purpose.rst_]: yes, a personal supplement - will need to check why but removed for now.  Thanks for the fixes!

File: doc/rst/source/modules.rst [hunk 5]
```diff
@@ -47,6 +47,7 @@ All modules are requested via a call to the :doc:`gmt` program.
    gmt2kml
    gmtconnect
    gmtconvert
+   gmtbinstats
```
[the remark this answers is not in the document]
Comment (PaulWessel) [reply doc/rst/source/modules.rst]: Done
